src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -615,6 +583,77 @@ RouterSequentialModifyExecScan(CustomScanState *node)
 }
 
 
+/*
+ * RouterSequentialModifyExecScan executes 0 or more modifications on a
+ * distributed table sequentially and stores them in custom scan's tuple
+ * store. Note that we also use this path for SELECT ... FOR UPDATE queries.
+ */
+static void
+RouterSequentialModifyExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	DistributedPlan *distributedPlan = scanState->distributedPlan;
+	bool hasReturning = distributedPlan->hasReturning;
+	Job *workerJob = distributedPlan->workerJob;
+	List *taskList = workerJob->taskList;
+	ListCell *taskCell = NULL;
+	bool multipleTasks = list_length(taskList) > 1;
+	EState *executorState = scanState->customScanState.ss.ps.state;
+	bool taskListRequires2PC = TaskListRequires2PC(taskList);
+	bool alwaysThrowErrorOnFailure = false;
+	CmdType operation = scanState->distributedPlan->operation;
+
+	Assert(!scanState->finishedRemoteScan);
+
+	/*
+	 * We could naturally handle function-based transactions (i.e. those using
+	 * PL/pgSQL or similar) by checking the type of queryDesc->dest, but some
+	 * customers already use functions that touch multiple shards from within
+	 * a function, so we'll ignore functions for now.
+	 */
+	if (IsTransactionBlock() || multipleTasks || taskListRequires2PC)
+	{
+		BeginOrContinueCoordinatedTransaction();
+
+		/*
+		 * Although using two phase commit protocol is an independent decision than
+		 * failing on any error, we prefer to couple them. Our motivation is that
+		 * the failures are rare, and we prefer to avoid marking placements invalid
+		 * in case of failures.
+		 *
+		 * For reference tables, we always set alwaysThrowErrorOnFailure since we
+		 * absolutely want to avoid marking any placements invalid.
+		 *
+		 * We also cannot handle failures when there is RETURNING and there are more
+		 * than one task to execute.
+		 */
+		if (taskListRequires2PC)
+		{
+			CoordinatedTransactionUse2PC();
+
+			alwaysThrowErrorOnFailure = true;
+		}
+		else if (multipleTasks && hasReturning)
+		{
+			alwaysThrowErrorOnFailure = true;
+		}
+	}
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		/*
+		 * Result is expected for SELECT ... FOR UPDATE queries as well.
+		 */
+		executorState->es_processed +=
+			ExecuteSingleModifyTask(scanState, task, operation,
+									alwaysThrowErrorOnFailure,
+									hasReturning || task->relationRowLockList != NIL);
```

Comment:
As a minor clean-up while we're doing this, it might be nice to declare an `expectResults` variable. It's also not obvious that the comment above applies to this line.